contracts/Portfolio.sol

```diff
@@ -444,44 +486,42 @@ abstract contract PortfolioVirtual is Objective {
             }
 
             deltaInputLessFee = deltaInput - iteration.feeAmount;
-
-            // This value should be used in `syncPool` if fees are re-invested into the pool.
-            nextIndependent =
-                liveIndependent + deltaInput.divWadDown(iteration.liquidity);
+            nextIndependentWad = liveIndependentWad + deltaInput;
 
             // This is a very critical piece of code!
-            // This value should be used in `syncPool` if fees are not re-invested.
+            // nextIndependentWadLessFee:
+            // This value should be used in `syncPool`.
             // The next independent amount is computed with the fee amount applied.
             // This means the lesser next independent reserve and dependent reserve
             // will pass the invariant.
+            //
+            // nextIndependent:
             // The fee amount has to be added to the reserve to re-invest it in the pool.
+            // So the next reserve should include the fee amount, since it was added to the reserves.
             // This will mean the independent reserve has more tokens than expected,
             // leading to a larger invariant.
-            nextIndependentLessFee = liveIndependent
-                + deltaInputLessFee.divWadDown(iteration.liquidity);
-            nextDependent =
-                liveDependent - deltaOutput.divWadDown(iteration.liquidity);
+            nextIndependentWadLessFee = liveIndependentWad + deltaInputLessFee;
+            nextDependentWad = liveDependentWad - deltaOutput;
         }
 
         // -=- Assert Invariant Passes -=- //
         {
             bool validInvariant;
             int256 nextInvariantWad;
 
-            // This is revisited depending on if fees are saved in claimable balances.
             if (_state.sell) {
                 (iteration.virtualX, iteration.virtualY) =
-                    (nextIndependentLessFee, nextDependent);
+                    (nextIndependentWadLessFee, nextDependentWad);
             } else {
                 (iteration.virtualX, iteration.virtualY) =
-                    (nextDependent, nextIndependentLessFee);
+                    (nextDependentWad, nextIndependentWadLessFee);
             }
 
             (validInvariant, nextInvariantWad) = checkInvariant(
                 args.poolId,
                 iteration.prevInvariant,
-                iteration.virtualX,
-                iteration.virtualY,
+                iteration.virtualX.divWadDown(iteration.liquidity), // Expects X per liquidity.
+                iteration.virtualY.divWadDown(iteration.liquidity), // Expects Y per liquidity.
```

Comment:
>  I'd still spend some more time thinking through this more and how it relates to the prevInvariant computed in beforeSwap.

Because you also round down when computing the `prevInvariant`, you can still get to the case where the reserves per liquidity end up being the same: https://github.com/spearbit-audits/review-primitive/issues/6#issuecomment-1512807314

Reply:
Rounds dependent reserve UP, keeping independent reserve rounded DOWN when doing the reserve per liquidity computations for the invariant check: https://github.com/primitivefinance/portfolio/commit/9f3768e1ae430645b774ba32d169db40a0f7c4d5

Reply:
cool, do these fuzz tests pass now?

Reply:
I added a fuzz test to fuzz swap input + outputs and if the swap goes through, check that the virtual reserves per liquidity did change. However, it's difficult to know if the fuzzing is finding enough successful swaps to actually do the check on.

Goal for us is to focus these fuzz tests in to cover this case, which we need https://github.com/primitivefinance/portfolio/pull/316 to merge in, since that's how we compute the exact swap amount out to get the invariant to stay unchanged (optimal) over a swap.